File: frontend/types/analysis.ts
export type JobStatus =
  | "pending"
  | "cloning"
  | "parsing"
  | "analyzing"
  | "completed"
  | "failed";

export interface AnalysisJob {
  id: string;
  repo_url: string | null;
  repo_name: string;
  status: JobStatus;
  error_message: string | null;
  total_files: number;
  total_tokens: number;
  created_at: string;
  updated_at: string;
}

export interface AnalysisJobDetail extends AnalysisJob {
  file_tree: Record<string, unknown> | null;
  ast_summary: AstSummary | null;
  migration_blueprint: MigrationBlueprint | null;
}

export interface AstSummary {
  total_files: number;
  total_tokens: number;
  files: AstFileEntry[];
}

export interface AstFileEntry {
  path: string;
  token_count: number;
  language?: string;
  imports?: string[];
  functions?: AstFunction[];
  classes?: AstClass[];
  error?: string;
}

export interface AstFunction {
  name: string;
  line: number;
  args: string[];
  is_async: boolean;
  decorators: string[];
  docstring: string | null;
}

export interface AstClass {
  name: string;
  line: number;
  bases: string[];
  docstring: string | null;
}

export interface BoundedContext {
  id: string;
  name: string;
  description: string;
  entities: string[];
  exposed_apis: string[];
  inbound_dependencies: string[];
  outbound_dependencies: string[];
  migration_phase: number;
}

export interface MigrationBlueprint {
  bounded_contexts: BoundedContext[];
  migration_phases: MigrationPhase[];
}

export interface MigrationPhase {
  phase: number;
  title: string;
  description: string;
  service_name: string;
  strangler_intercepts: string[];
}

export const JOB_STATUS_STEPS: Record<
  JobStatus,
  { label: string; progress: number }
> = {
  pending: { label: "Queued", progress: 5 },
  cloning: { label: "Cloning repository…", progress: 20 },
  parsing: { label: "Parsing AST…", progress: 50 },
  analyzing: { label: "Analyzing with AI…", progress: 75 },
  completed: { label: "Analysis complete", progress: 100 },
  failed: { label: "Failed", progress: 0 },
};

File: backend/app/schemas/analysis.py
import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, HttpUrl

from app.db.models.analysis import JobStatus


class AnalysisJobCreate(BaseModel):
    repo_url: HttpUrl
    github_token: str | None = None


class AnalysisJobResponse(BaseModel):
    id: uuid.UUID
    repo_url: str | None
    repo_name: str
    status: JobStatus
    error_message: str | None
    total_files: int
    total_tokens: int
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class AnalysisJobDetail(AnalysisJobResponse):
    file_tree: dict[str, Any] | None
    ast_summary: dict[str, Any] | None
    migration_blueprint: dict[str, Any] | None


class HealthResponse(BaseModel):
    status: str
    environment: str
    version: str = "0.1.0"
